tests/python/relay/test_annotated_regions.py

```diff
@@ -56,7 +56,7 @@ def test_region_set_creator_diamond():
         'test_target',
         [cb_1],
         [cb_1, O_1, ce_1, ce_2],
-        [ce_1, ce_2],
```

Comment:
@mbaret @comaniac `ce_1` and `ce_2` are duplicates here. I had to make this change to this test to get it to pass after my fix. Is it important that AnnotatedRegions can have duplicate outputs? Is there a use case for annotatedRegions outside of the MergeCompilerRegions/PartitionGraph flow?

Reply:
This is a valuable question. For the use case of partition graph pass, duplicated outputs are definitely not required. However, for the semantic of `AnnotatedRegionSet` itself, it seems weird to find 2 boundary `compiler_ends` in node list but only one of them is marked as an output. Accordingly, I have 3 proposals:

A. Let `AnnotatedRegionSet` maintain unique outputs as this PR. We may add some comments mentioning that if there are duplicated `compiler_end`s in a region, we will randomly mark one of them as an output.

B. `AnnotatedRegionSet` still keeps all outputs, but provides another API (e.g., `GetUniqueOutputs()`) that contains the de-duplication logic in this PR. `GetUniqueOutputs` returns another private class member `unique_outs_` which will be updated every time `outs_` is updated. It also means that `outs_` must be a private member as well and can only be updated via APIs.

C. `AnnotatedRegionSet` does not care about duplicated outputs. It will mark all `compiler_end`s at boundary as outputs. In this case, `PartitionGraph` needs to maintain a cache for region outputs.

cc @trevor-m @zhiics @mbaret @manupa-arm

Reply:
Thanks @comaniac for the review and comments! I like your proposed solutions B and C. B seems to be the cleanest to implement while C helps keep the passes not codependent on each other as much. I would be happy to implement any of those, what do others think?

Reply:
We introduced a map #5310 to cache the region output, can we reuse it?

Reply:
`region_args` introduced in that PR is for region inputs. We may still need another `region_outs` map, and this fits to the proposal C.

Reply:
Apologies for the delayed response, we've been on holiday in the UK. I think C would be best as `AnnotatedRegionSet` probably shouldn't care about de-duplication logic - that's a specific requirement of `PartitionGraph`.

Reply:
Thanks all, I've implemented proposal C with the latest commits.